src/isdimm_vpd_parser.cpp

```diff
@@ -0,0 +1,366 @@
+constexpr auto SPD_JEDEC_DDR4_DRAM_PRI_PACKAGE_OFFSET = 6;
+constexpr auto SPD_JEDEC_DDR4_DRAM_MODULE_ORG_OFFSET = 12;
+
+// Lookup tables
+const types::PartNumberFreqMap pnFreqFnMap = {
+    {std::make_tuple("8421000", 6), "78P4191"},
+    {std::make_tuple("8421008", 6), "78P4192"},
+    {std::make_tuple("8529000", 6), "78P4197"},
+    {std::make_tuple("8529008", 6), "78P4198"},
+    {std::make_tuple("8529928", 6), "78P4199"},
+    {std::make_tuple("8529B28", 6), "78P4200"},
+    {std::make_tuple("8631928", 6), "78P6925"},
+    {std::make_tuple("8529000", 5), "78P7317"},
+    {std::make_tuple("8529008", 5), "78P7318"},
+    {std::make_tuple("8631008", 5), "78P6815"}};
+
+const types::PartNumberCCINMap pnCCINMap = {
+    {"78P4191", "324D"}, {"78P4192", "324E"}, {"78P4197", "324E"},
+    {"78P4198", "324F"}, {"78P4199", "325A"}, {"78P4200", "324C"},
+    {"78P6925", "32BC"}, {"78P7317", "331A"}, {"78P7318", "331F"},
+    {"78P6815", "32BB"}};
+
+// Helper functions
+
+/**
+ * @brief Calculate Part Number.
+ *
+ * Calculates Part number for DDR4 Dimm as defined and returns
+ * value if above 0 else null option.
+ *
+ * @param[in] i_iterator - iterator to buffer containing SPD
+ * @return Null when error or string
+ */
+std::optional<std::string_view>
+    calculatePartNumber(const types::BinaryVector::const_iterator& i_iterator)
+{
+    char l_tmpPN[constants::PART_NUM_LEN + 1] = {'\0'};
+    if (sprintf(l_tmpPN, "%02X%02X%02X%X",
+                i_iterator[SPD_JEDEC_DDR4_SDRAM_DENSITY_BANK_OFFSET],
+                i_iterator[SPD_JEDEC_DDR4_SDRAM_ADDR_OFFSET],
+                i_iterator[SPD_JEDEC_DDR4_DRAM_PRI_PACKAGE_OFFSET],
+                i_iterator[SPD_JEDEC_DDR4_DRAM_MODULE_ORG_OFFSET] & 0x0F) <= 0)
+    {
+        return std::nullopt; // Error occurred
+    }
+    return std::string_view(l_tmpPN);
+}
+
+/**
+ * @brief Get FRU Number.
+ *
+ * Gets FRU number for DDR4 Dimm as defined and returns
+ * value if found else null option.
+ *
+ * @param[in] i_partNumber - Part number
+ * @param[in] i_iterator - iterator to buffer containing SPD
+ * @param[in] i_freqMap - FRU map
+ * @return Null when error or String
+ */
+std::optional<std::string_view>
+    getFruNumber(const std::string& i_partNumber,
+                 const types::BinaryVector::const_iterator& i_iterator,
+                 const types::PartNumberFreqMap& i_freqMap)
+{
+    uint8_t l_mtbUnits = i_iterator[constants::SPD_BYTE_18] &
+                         constants::SPD_BYTE_MASK;
+    auto it = i_freqMap.find({i_partNumber, l_mtbUnits});
+    if (it != i_freqMap.end())
+    {
+        return it->second;
+    }
+    return std::nullopt; // Not found
+}
+
+/**
+ * @brief Get CCIN
+ *
+ * Gets CCIN from fru number returns value if exists else default string.
+ *
+ * @param[in] i_fruNumber - FRU number
+ * @param[in] i_ccinMap - CCIN map from FRU number
+ * @return CCIN string or default value when not found
+ */
+std::string_view getCCIN(const std::string& i_fruNumber,
```

Comment:
Why global functions?

Reply:
Thought can be useful for DDR5/other too with a flag later, when it's available.

Reply:
No, Please avoid using global functions.